### ISSUE TITLE
Language iteration cycle estimate grid

### ISSUE THREAD

**Issue by pctj101** [apex/up#439]
TLDR; want to know how long it takes to deploy an update for each supported language

----

Part of the annoyance of lambda is how a simple script-function updates so quickly (wow so cool) when there are no dependencies, then quickly proceeds into a giant multi-megabyte upload when adding in all node_modules... or golang dependencies for that matter (wow so not cool).  This is especially painful when just adding commas and needing to wait for uploads.

For someone like me who wants fast iteration cycles, it would be great for the Apex Up readme to simply say "Typically, by the time you add in aws sdk, url parsers, etc..." (whatever that mythical bundle is), this is what you can expect your "up deploy" time/sizes to be like (assuming x mbps internet connection).

I know... this is far from perfect.  I just dream of a place where someone explains how to make lambda deploys fast.  When I pay $x for a subscription (which I did!), I kind of want that guidance as my "instant refund". :)

Is it fair to say that one should "figure it out own their own" because lambda already has these inherent "times"?  Yes... sorta.  Golang/Rust is supported here, and typically there's no mental basis for that.

So do I want to try golang/rust if I know the deploy is 5x faster? Yes!  What if it's 10x slow? No!  Or... where's that line?

**Comment by pctj101** [comment]
I suppose I could just... deploy all the examples one at a time and time it?  Well... not really... I suppose this is harder than it sounds.

**Comment by tj** [comment]
It'll vary a lot depending on the project and your connection speed. My real-world go apps deploy in ~10-15s from a Starbucks, but as you mention if you're using something like Node with a lot of node_modules it can become large quickly. 

In those cases it's definitely more ideal to deploy from CI so you can utilize their network. Eventually I'd like to support Docker and/or diffs etc, but for now just simple zips is it. 

Which language(s) are you planning on using?

**Comment by pctj101** [comment]
On my side it's javascript and golang.

---

A simple express app w/ aws-sdk
```
  "dependencies": {
    "async": "^2.6.0",
    "aws-sdk": "^2.156.0",
    "aws-serverless-express": "^3.0.2",
    "body-parser": "^1.18.2",
    "compression": "^1.7.1",
    "cors": "^2.8.4",
    "express": "^4.15.3",
    "express-favicon": "^2.0.0",
    "pug": "^2.0.0-rc.4"
  }
```
deploy time
```
     hook: build (9ms)
     build: 4,663 files, 13 MB (3.423s)
     deploy: complete (10.003s)
```
----

Deploying gh-polls which appears to use aws-sdk-go, pat -- close enough
```
     hook: build (24ms)
     build: 23 files, 4.4 MB (530ms)
     deploy: complete (2.833s)
     hook: clean (18ms)
```

----

so there's definitely a good reason to run away from the billions of node_module files.

**Comment by tj** [comment]
Haha yep definitely, I wrote tj/node-prune to help a bit there, but it doesn’t completely solve the problem. Sometimes you can get away with using rollup/webpack as well. 

Going to close for now, I think there are too many variables, plus medium-to-large companies typically use CI where the network is much better as well.

**Comment by pctj101** [comment]
Wow built in CI.  Yeah whatever the flavor is, it would be quite cool to take another integration step off my mind.  Anyways, your thoughts are well received!

**Comment by tj** [comment]
Agreed! It's definitely on the list, they just announced some great CI features too so I'd like to streamline that
